dataclients/kubernetes/crd.md

```diff
@@ -0,0 +1,634 @@
+complex.example.org.
+
+We might want to specify a redirect to our login service.
+
+Maybe some day we want to migrate from /api to /, in our backend. Old services should be able to use /api, which they can later change to /.
+
+```yaml
+apiVersion: zalando.org/v1
+kind: RouteGroup
+metadata:
+  name: my-routes
+spec:
+  hosts:
+  - www.complex.example.org
+  - complex.example.org
+  backends:
+  - name: my-service
+    type: service
+    serviceName: my-service-v1
+    servicePort: 80
+    default: true
+  - name: redirect
+    type: shunt
+  - name: migration
+    type: loopback
+  routes:
+  - path: /
+  - path: /api
+    backend: migration
+    filters:
+    - modPath("/api", "/")
+  - path: /login
+    method: get
+    backend: redirect
+    filters:
+    - redirectTo(308, "https://login.example.org/)
+```
+
+
+#### Complex routes with ratelimits based on tokens
+
+A complex route case could specify different ratelimits for POST
+request to /api/resource for clients.  Clients that have JWT/OAuth2
+Tokens from issuer https://accounts.google.com with email
+"important@example.org" get 20/equest per minute, other clients with
+Token with issuer https://accounts.google.com or
+https://accounts.github.com get only 2.
+
+Additionally for all other HTTP methods ratelimit for each client 100
+requests per minute based on "Authorization" header and make sure the
+Token is valid and has the scopes to read and list the resource.
+
+```yaml
+apiVersion: zalando.org/v1
+kind: RouteGroup
+metadata:
+  name: api
+spec:
+  hosts:
+  - api.example.org
+  - api.service.example.net
+  backends:
+  - name: api-svc
+    type: service
+    serviceName: api-service-v1
+    servicePort: 80
+    default: true
+  routes:
+  - path: /api/resource
+    method: post
+    filters:
+    - ratelimit(20, "1m")
+    - oauthTokeninfoAllKV("iss", "https://accounts.google.com", "email", "important@example.org")
+    predicates:
+    - JWTPayloadAllKV("iss", "https://accounts.google.com", "email", "skipper-router@googlegroups.com")
+  - path: /api/resource
+    method: post
+    filters:
+    - ratelimit(2, "1m")
+    - oauthTokeninfoAnyKV("iss", "https://accounts.google.com", "iss", "https://accounts.github.com")
+  - path: /api/resource
+    filters:
+    - clientRatelimit(100, "1m", "Authorization")
+    - oauthTokeninfoAllScope("read.resource", "list.resource")
+```
+
+#### Traffic switching
+
+If we want to traffic switch, we want to make sure all path endpoints
+we have configured are switched at the same time and get the same
+traffic split applied. To not be ambiguous with multiple backends, we
+need to set `default: true` for both backends we traffic switched.
+
+```yaml
+apiVersion: zalando.org/v1
+kind: RouteGroup
+metadata:
+  name: my-routes
+spec:
+  hosts:
+  - api.example.org
+  - api.service.example.net
+  backends:
+  - name: api-svc
+    type: service
+    serviceName: api-service-v1
+    servicePort: 80
+    default: true
+    weight: 80
```

Comment:
Copying my comments from the other proposal document.

The semantics of this are very unclear and I think this needs to be split. Here, we define the implementation of the backend, whether it's a default or not, and its traffic weight in one declaration, but these are completely different things. For example, I might want to define three different services, have two of them be default ones with traffic weights 70 and 30, but then want to use all three in a dedicated route with different traffic weights, and how would I do that? Or, what exactly is going to happen if one of the service is no longer marked as default but still has a traffic weight?

I propose to change this as follows:

1) Have one section to define the backends:

       backends:
         - name: api-svc-v1
           type: service
           serviceName: api-service-v1
           servicePort: 80
         - name: api-svc-v2
           type: service
           serviceName: api-service-v2
           servicePort: 80

2) Then add another section to configure the default backends. This is a simple list of backend names and their traffic weights. If unset, and there's just one backend in the backend list, make it default to that backend only:

       defaultBackends:
         - name: api-svc-v1
           weight: 70
         - name: api-svc-v2
           weight: 30

3) Change the per-route backend to the same structure as the default backends (since the users should have freedom to define custom backends per-route):
    
       - path: /api
         backends:
          - name: api-svc-v2
            weight: 100

Reply:
So you suggest:

```yaml
apiVersion: zalando.org/v1
kind: RouteGroup
spec:
  hosts:
  - <string>
  backends:
  - <backend>
  default_backends: optional
  - <default_backend>
  routes:
  - <route>

<backend>
  name: <string>
  type: <string> that is one of "service|shunt|loopback|lb|url"
  url: <string> required for type=url
  algorithm: <string> optional, only user for type=lb
  endpoints: <stringarray> required for type=lb
  serviceName: <string> required for type=service
  servicePort: <number> required for type=service

<default_backend>
  default: <bool>
  weight: <number>

<route>
  path: <string>
  exactPath: <bool> optional
  backend: optional, overrides defaults
  - <backend_ref>
  filters: <stringarray>, optional
  predicates: <stringarray>, optional
  method: <string>, optional, one of the HTTP methods "GET|HEAD|PATCH|POST|PUT|DELETE|CONNECT|OPTIONS|TRACE", defaults to all

<backend_ref>
  - backend_name: <string>
    weight: <number>, optional


<string>  an arbitrary string
<stringarray> is an array of <string>
<number> is an unsigned integer, example 80
<bool> is a either true or false
```

Reply:
`default_backends` and `route.backends` should use the same structure that's a list of `backend_ref`.

Reply:
@szuecs Please fix the casing to `camelCase` (default_backends, backend_name, etc).

How does `path_regexp` work? Does it override `path`/`path_subtree`, or can you use it at the same time as those?

Reply:
Also, `route.backend` -> `route.backends`.

Reply:
Thanks will push in a bit

Reply:
fixed all the things